code/ui/blocks/src/blocks/Canvas.tsx

```diff
@@ -36,31 +115,120 @@ const usePreviewProps = (
   const stories = useStories(storyIds, docsContext);
   const isLoading = stories.some((s) => !s);
   const sourceProps = useSourceProps(
-    mdxSource ? { code: decodeURI(mdxSource) } : { ids: storyIds },
+    mdxSource ? { code: decodeURI(mdxSource), of: props.of } : { ids: storyIds, of: props.of },
     docsContext,
     sourceContext
   );
   if (withSource === SourceState.NONE) {
     return { isLoading, previewProps: props };
   }
 
+  // if the user has specified a layout prop, use that...
+  let layout = layoutProp;
+  // ...otherwise, try to infer it from the children 'parameters' prop
+  Children.forEach(children, (child) => {
+    if (layout) {
+      return;
+    }
+    layout = (child as ReactElement)?.props?.parameters?.layout;
+  });
+  // ...otherwise, try to infer it from the story parameters
+  stories.forEach((story) => {
+    if (layout || !story) {
+      return;
+    }
+    layout = story?.parameters.layout ?? story.parameters.docs?.canvas?.layout;
+  });
+
   return {
     isLoading,
     previewProps: {
       ...props, // pass through columns etc.
+      layout: layout ?? 'padded',
       withSource: sourceProps,
       isExpanded: (withSource || sourceProps.state) === SourceState.OPEN,
     },
   };
 };
 
-export const Canvas: FC<CanvasProps> = (props) => {
+export const Canvas: FC<CanvasProps & DeprecatedCanvasProps> = (props) => {
   const docsContext = useContext(DocsContext);
   const sourceContext = useContext(SourceContext);
-  const { isLoading, previewProps } = usePreviewProps(props, docsContext, sourceContext);
-  const { children } = props;
+  const { children, of, source } = props;
+  const { isLoading, previewProps } = useDeprecatedPreviewProps(props, docsContext, sourceContext);
+
+  let story;
+  let sourceProps;
+  /**
+   * useOf and useSourceProps will throw if they can't find the story, in the scenario where
+   * the doc is unattached (no primary story) and 'of' is undefined.
+   * That scenario is valid in the deprecated API, where children is used as story refs rather than 'of'.
+   * So if children is passed we allow the error to be swallowed and we'll use them instead.
+   * We use two separate try blocks to not break the rules of hooks.
+   */
+  try {
+    ({ story } = useOf(of || 'story', ['story']));
+  } catch (error) {
+    if (!children) {
+      throw error;
+    }
```

Comment:
I guess to not break the rules of hooks we would have to not throw here but instead "save" the error until after the second hook has run, then throw it.

I think either that, or the original is better, this halfway house is sort of pointless (it won't actually mean the hooks will do the right thing, right?).

Up to you; I think we can safely "break" the rules of hooks in this case given props are unlikely to change in a way that would violate them.

Reply:
oh you're right, that was a blunder from me. I'll fix it better, but I also think you're right that it's very unlikely we'll ever run into this.